### DISCUSSION TITLE
[next-auth][error][CLIENT_FETCH_ERROR] - How to Build a Fullstack App with Next.js, Prisma, and Vercel Postgres

### DISCUSSION THREAD

**Discussion by prtpj1** [vercel/next.js#60536] [Help] [answered]
### Summary

Hi everyone o/

I'm following this tutorial to learn about framework NextJS.
But I got a issue with the authentication.

All envs are ok (the console.log showed to me it was passed correctly)

I looked at google and stackoverflow and the only fix I found was to add a e.PreventDefault() to the Siging button, but on this tutorial is using a <Link>

```
<Link href='/api/auth/signin' legacyBehavior>
          <a data-active={isActive('/signup')}>Log in</a>
</Link>
```

I dont know how or if its possible to add e.PreventDefault() to this (I believe isnt possible). So I couldnt use that fix

I found too another person with the same issue, but there isnt any answer there.
The repository:
https://github.com/prtpj1/blogr-nextjs-prisma

Thx to who can help with this 

### Additional information

```js
npm run dev

> hello-next@1.0.0 dev
> next

   ▲ Next.js 14.0.4
   - Local:        http://localhost:3000
   - Environments: .env

 ✓ Ready in 1940ms
 ○ Compiling /drafts ...
 ✓ Compiled /drafts in 2.2s (485 modules)
 ✓ Compiled /api/auth/[...nextauth] in 98ms (144 modules)
 ✓ Compiled /create in 197ms (508 modules)
 ✓ Compiled /api/post in 43ms (148 modules)
[next-auth][error][CLIENT_FETCH_ERROR] 
https://next-auth.js.org/errors#client_fetch_error undefined {
  error: {},
  url: 'http://localhost:3000/api/auth/session',
  message: undefined
}
```

**Comment by 808vita** [comment]
two things :
1 - i think `<a>` tag nested inside `<Link>` is wrong you might receive errors for this . Why not directly add the link text within <Link> tag?

```
<Link href='/api/auth/signin' data-active={isActive('/signup')}>
 Log in
</Link>
```

2- in your code you have used `getSession` but try out `getServerSession` instead.
For server side code I would prefer to use getServerSession .
[index.ts](https://github.com/prtpj1/blogr-nextjs-prisma/blob/main/pages/api/post/index.ts)


I have used nextjs with next auth & mongodb adaptor multiple times and it always worked ; but have not tried out prisma .
[glance through this ](https://github.com/808vita/nextjs-blog-with-auth) - maybe this helps you (but I used MongoDBAdapterr ).

**Reply by prtpj1** [reply]
Hi 808vita, thx for the reply.

About the 2 things:
1 - I got an error and there was a link to vercel pages how to fix it.
https://nextjs.org/docs/messages/invalid-new-link-with-extra-anchor
Why I did that way? I was just following the tutorial. Usually I dont change tutorials until finish it and everything working fine. Then I start to make changes to see what happens and what I can improve.
![Screenshot_342](https://github.com/vercel/next.js/assets/91347088/92963cc0-bc9b-4f30-9119-2fe8fa613961)

2 - I will try your suggestion, but as I said, I was just following the tutorial 😄

**Reply by 808vita** [reply]
Fair enough, but there is one major difference the tutorial used earlier versions of multiple dependencies. For example you have used next14 but the sample repo uses next 12 ; multiple breaking changes were introduced in multiple dependencies over time.
Please compare package.json files.

Things can get really confusing. 
You will have a better learning experience if you replicate the tutorial with same versions of dependencies and after it works as required you can try to upgrade one dependency at a time. Easier to track , fix bugs and overall less confusing (personal opinion :smile: )

**Reply by prtpj1** [reply]
I tried, but didnt work.
![Screenshot_343](https://github.com/vercel/next.js/assets/91347088/2f021124-72c7-4fcb-8302-0f4363d1a318)

I'm getting this error after the changes:
```
npm run dev                                                                                                                                                                    ⏎

> hello-next@1.0.0 dev
> next

   ▲ Next.js 14.0.4
   - Local:        http://localhost:3000
   - Environments: .env

 ✓ Ready in 1405ms
 ✓ Compiled /_error in 476ms (235 modules)
 ⚠ Fast Refresh had to perform a full reload. Read more: https://nextjs.org/docs/messages/fast-refresh-reload
 ✓ Compiled /drafts in 312ms (489 modules)
 ⚠ Fast Refresh had to perform a full reload. Read more: https://nextjs.org/docs/messages/fast-refresh-reload
 ✓ Compiled /api/auth/[...nextauth] in 105ms (146 modules)
 ✓ Compiled /create in 180ms (512 modules)
 ✓ Compiled /api/post in 96ms (150 modules)
 ⨯ Error: Invariant: headers() expects to have requestAsyncStorage, none available.
    at headers (/home/prtpj/Trybe_U/Projetos/tutoriais/react/blogr-nextjs-prisma/node_modules/next/dist/client/components/headers.js:42:15)
    at getServerSession (/home/prtpj/Trybe_U/Projetos/tutoriais/react/blogr-nextjs-prisma/node_modules/next-auth/next/index.js:139:35)
    at handle (webpack-internal:///(api)/./pages/api/post/index.ts:12:86)
    at /home/prtpj/Trybe_U/Projetos/tutoriais/react/blogr-nextjs-prisma/node_modules/next/dist/compiled/next-server/pages-api.runtime.dev.js:21:3039
    at /home/prtpj/Trybe_U/Projetos/tutoriais/react/blogr-nextjs-prisma/node_modules/next/dist/server/lib/trace/tracer.js:133:36
    at NoopContextManager.with (/home/prtpj/Trybe_U/Projetos/tutoriais/react/blogr-nextjs-prisma/node_modules/next/dist/compiled/@opentelemetry/api/index.js:1:7062)
    at ContextAPI.with (/home/prtpj/Trybe_U/Projetos/tutoriais/react/blogr-nextjs-prisma/node_modules/next/dist/compiled/@opentelemetry/api/index.js:1:518)
    at NoopTracer.startActiveSpan (/home/prtpj/Trybe_U/Projetos/tutoriais/react/blogr-nextjs-prisma/node_modules/next/dist/compiled/@opentelemetry/api/index.js:1:18093)
    at ProxyTracer.startActiveSpan (/home/prtpj/Trybe_U/Projetos/tutoriais/react/blogr-nextjs-prisma/node_modules/next/dist/compiled/@opentelemetry/api/index.js:1:18854)
    at /home/prtpj/Trybe_U/Projetos/tutoriais/react/blogr-nextjs-prisma/node_modules/next/dist/server/lib/trace/tracer.js:122:103
    at NoopContextManager.with (/home/prtpj/Trybe_U/Projetos/tutoriais/react/blogr-nextjs-prisma/node_modules/next/dist/compiled/@opentelemetry/api/index.js:1:7062)
    at ContextAPI.with (/home/prtpj/Trybe_U/Projetos/tutoriais/react/blogr-nextjs-prisma/node_modules/next/dist/compiled/@opentelemetry/api/index.js:1:518)
    at NextTracerImpl.trace (/home/prtpj/Trybe_U/Projetos/tutoriais/react/blogr-nextjs-prisma/node_modules/next/dist/server/lib/trace/tracer.js:122:28)
    at K (/home/prtpj/Trybe_U/Projetos/tutoriais/react/blogr-nextjs-prisma/node_modules/next/dist/compiled/next-server/pages-api.runtime.dev.js:21:2970)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  page: '/api/post'
}
```

**Reply by 808vita** [reply]
for getServerSession i usually pass req,res and authoptions
[doc link](https://next-auth.js.org/configuration/nextjs)

**Reply by prtpj1** [reply]
Hi 808vita
I tried to follow the link you sent, but probably its above me now. 

My codes after follow(failed) instructions:
**Drafts.tsx**
```
export const getServerSideProps: GetServerSideProps = async ({ req, res }) => {
    const session = await getServerSession(req);

    if (!session) {
        res.statusCode = 403;
        return { props: { drafts: [] } };
    }

    const drafts = await prisma.post.findMany({
        where: {
            author: { email: session.user.email },
            published: false,
        },
        include: {
            author: {
                select: { name: true },
            },
        },
    });
    return {
        props: { drafts },
    };
};
```

**index.ts**
```
export default async function handle(req: any, res: any) {
    const { title, content } = req.body;

    const session: Session = await getServerSession(req);
    if (!session || !session.user.email) {
        res.status(401).json({ error: 'You must be logged in to create a post.' });
        return;
    }

    const result = await prisma.post.create({
        data: {
            title: title,
            content: content,
            author: { connect: { email: session?.user?.email } },
        },
    });
    res.json(result);
}
```

**[...nextauth].ts**
```
export const authHandler: NextAuthOptions = {
    providers: [
        GitHubProvider({
            clientId: process.env.GITHUB_ID,
            clientSecret: process.env.GITHUB_SECRET,
        }),
    ],
    adapter: PrismaAdapter(prisma),
    secret: process.env.SECRET,
}

export default NextAuth(authHandler);
```

There is a way to contact vercel to send a feedback about this tutorial issue?
To who is learning its a little challenge

I will look for another nextjs tutorials trying to learn more about it

**Reply by 808vita** [reply, marked as answer]
You are almost there , you got this.

**In `index.ts` :**
Make sure to import the authopitons :
` import { authHandler } from "../auth/[...nextauth]";`

And then :
instead of `await getServerSession(req)`

```
 /**
   *  get session for api route protection
   */
  const session = await getServerSession(req, res, authHandler);
```

And yes you should raise an issue in the sample repo , probably maintained by prisma (not sure?)

**Reply by 808vita** [reply]
Some folks have already raised an issue [here](https://github.com/prisma/blogr-nextjs-prisma/issues)

**Reply by prtpj1** [reply]
Hi 808vita

Thx a lot for your help and patience

Now its working and I can finish the tutorial ( I hope 😋 )

I had to add too, it to Draft.tsx
`const session = await getServerSession(req, res, authHandler);`
